source/MissionAction.cpp

```diff
@@ -384,6 +402,8 @@ MissionAction MissionAction::Instantiate(map<string, string> &subs, int jumps, i
 	MissionAction result;
 	result.trigger = trigger;
 	result.system = system;
+	// Convert any "distance" specifiers into "near <system>" specifiers.
+	result.systemFilter = systemFilter.SetOrigin(GameData::Planets().Get(subs["<origin>"])->GetSystem());
```

Comment:
This won't work if the origin is a ship you're boarding (in which case `"<origin>"` is the name of the ship, not the name of a system). It might be necessary to just pass the current system to `MissionAction::Instantiate()`.